src/libraries/System.Net.Connections/src/System/Net/Connections/Connection.cs

```diff
@@ -0,0 +1,288 @@
+        {
+            if (pipe == null) throw new ArgumentNullException(nameof(pipe));
+            return new ConnectionFromPipe(pipe, disposable, properties, localEndPoint, remoteEndPoint);
+        }
+
+        private sealed class ConnectionFromStream : Connection, IConnectionProperties
+        {
+            private new Stream? _stream;
+            private IConnectionProperties? _properties;
+
+            public override IConnectionProperties ConnectionProperties => _properties ?? this;
+
+            public override EndPoint? LocalEndPoint { get; }
+
+            public override EndPoint? RemoteEndPoint { get; }
+
+            public ConnectionFromStream(Stream stream, IConnectionProperties? properties, EndPoint? localEndPoint, EndPoint? remoteEndPoint)
+            {
+                _stream = stream;
+                _properties = properties;
+                LocalEndPoint = localEndPoint;
+                RemoteEndPoint = remoteEndPoint;
+            }
+
+            protected override Stream CreateStream() => _stream ?? throw new ObjectDisposedException(nameof(Connection));
+
+            protected override async ValueTask CloseAsyncCore(ConnectionCloseMethod method, CancellationToken cancellationToken)
+            {
+                if (_stream == null)
+                {
+                    return;
+                }
+
+                if (method == ConnectionCloseMethod.GracefulShutdown)
+                {
+                    await _stream.FlushAsync(cancellationToken).ConfigureAwait(false);
+                }
+
+                await _stream.DisposeAsync().ConfigureAwait(false);
+
+                _stream = null;
+            }
+
+            bool IConnectionProperties.TryGet(Type propertyKey, [NotNullWhen(true)] out object? property)
+            {
+                property = null;
+                return false;
+            }
+        }
+
+        private sealed class ConnectionFromPipe : Connection, IConnectionProperties
+        {
+            private new IDuplexPipe? _pipe;
+            private object? _disposable;
+            private IConnectionProperties? _properties;
+
+            public override IConnectionProperties ConnectionProperties => _properties ?? this;
+
+            public override EndPoint? LocalEndPoint { get; }
+
+            public override EndPoint? RemoteEndPoint { get; }
+
+            public ConnectionFromPipe(IDuplexPipe pipe, object? disposable, IConnectionProperties? properties, EndPoint? localEndPoint, EndPoint? remoteEndPoint)
+            {
+                Debug.Assert(disposable == null || disposable is IDisposable || disposable is IAsyncDisposable);
+
+                _pipe = pipe;
+                _disposable = disposable;
+                _properties = properties;
+                LocalEndPoint = localEndPoint;
```

Comment:
Why do we need this? For logging?

Reply:
Yea. This implementation is for convenience, normal implementations will get local endpoint from e.g. a socket.